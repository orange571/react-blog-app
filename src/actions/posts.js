import uuid from 'uuid';
import database from '../firebase/firebase';
import moment from 'moment'

// ADD_POST
export const addPost = (post) => ({
  type: 'ADD_POST',
  post
});

export const startAddPost = (postData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      body = '',
      createdAt = 0,
      isReadable = false,
      author = '',
    } = postData;
    const post = { title, body, createdAt, isReadable, author, uid };

    return database.ref(`posts`).push(post).then((ref) => {
      const postId = ref.key;
      database.ref(`users/${uid}/posts`).update({[postId] : isReadable}).then(()=>{
        if(isReadable){
          database.ref(`isPublic`).push({id: postId})
        };
      });
      dispatch(addPost({
        id: ref.key,
        ...post
      }));
    });
  };
};

// REMOVE_POST
export const removePost = ({ id } = {}) => ({
  type: 'REMOVE_POST',
  id
});

export const startRemovePost = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`posts/${id}`).remove().then(() => {
      //REMOVE FROM PUBLIC LIST IF WAS THERE BEFORE
      database.ref(`isPublic`).orderByChild("id").equalTo(id).once("value",snapshot => {
          const wasPublic = snapshot.val();
          if (!!wasPublic){
            const publicPostId = Object.keys(wasPublic)[0];
            database.ref(`isPublic/${publicPostId}`).remove().then(() => {
              //REMOVE FROM USER List
              database.ref(`users/${uid}/posts/${id}`).remove().then(()=>console.log("success remove from user"))
              .catch(error => console.log(error));
            }).catch(error => console.log(error));
          } else {
            //REMOVE FROM USER List
            database.ref(`users/${uid}/posts/${id}`).remove().then(()=>console.log("success remove from user"))
            .catch(error => console.log(error));
          }
      });
      dispatch(removePost({ id }));
    });
  };
};

// EDIT_POST
export const editPost = (id, updates) => ({
  type: 'EDIT_POST',
  id,
  updates
});

export const startEditPost = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`posts/${id}`).update(updates).then(() => {
      //ADD TO PUBLIC LIST IF WAS NOT THERE ALREADY
      if(updates.isReadable){
        database.ref(`isPublic`).orderByChild("id").equalTo(id).once("value",snapshot => {
            const wasPublic = snapshot.val();
            if (!wasPublic){
              database.ref(`isPublic`).push({id: id})
            }
        });
        //REMOVE FROM PUBLIC LIST IF WAS THERE BEFORE
      } else {
        database.ref(`isPublic`).orderByChild("id").equalTo(id).once("value",snapshot => {
            const wasPublic = snapshot.val();
            const publicPostId = Object.keys(wasPublic)[0];
            if (!!wasPublic){
              database.ref(`isPublic/${publicPostId}`).remove();
            }
        });
      }
      database.ref(`users/${uid}/posts`).update({[id]: updates.isReadable});
      dispatch(editPost(id, updates));
    });
  };
};

// SET_POSTS
export const setPosts = (posts) => ({
  type: 'SET_POSTS',
  posts
});

export const startSetPosts = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts`).once('value').then((snapshot) => {
      if(!snapshot.val()) {
        let posts = [];
        dispatch(setPosts(posts));
      } else {
        return new Promise((resolve, reject) => {
          const postIds = [];
          snapshot.forEach((childSnapshot) => {
            postIds.push(childSnapshot.key);
          });
          let posts = [];
          postIds.forEach((postId, index)=>{
            database.ref(`posts/${postId}`).once('value').then((postSnapshot) => {
              posts.push({
                id: postSnapshot.key,
                ...postSnapshot.val()
              });
              if (index === postIds.length-1) {
                resolve(posts);
              }
            }).catch((error) => {
              console.log("error", error);
              reject(error);
            });
          });
        }).then((posts)=>{
          dispatch(setPosts(posts));
        }).catch(error => {
          console.log(error)
        });
      }
    }).catch((error)=>{
      console.log('error', error);
    });
  };
};
