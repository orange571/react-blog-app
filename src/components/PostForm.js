import React from 'react';
import moment from 'moment';

export default class PostForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      title: props.post ? props.post.title : '',
      body: props.post ? props.post.body : '',
      createdAt: props.post ? moment(props.post.createdAt) : moment(),
      author: props.post ? props.post.author: props.displayName,
      error: ''
    };
  }
  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({ title }));
  };
  onBodyChange = (e) => {
    const body = e.target.value;
    this.setState(() => ({ body }));
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.title || !this.state.body) {
      this.setState(() => ({ error: 'Please provide a title and post body.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        title: this.state.title,
        createdAt: this.state.createdAt.valueOf(),
        body: this.state.body,
        author: this.state.author.valueOf(),
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="Title"
          autoFocus
          className="text-input"
          value={this.state.title}
          onChange={this.onTitleChange}
        />
        <textarea
          placeholder="Add a body for your post"
          className="textarea"
          value={this.state.body}
          onChange={this.onBodyChange}
        >
        </textarea>
        <div className="form__button-container">
          <button type="submit" className="button">Save Post</button>
          {this.props.onRemove && (<button className="button button--secondary" onClick={this.props.onRemove}>Remove Post</button>)}
        </div>
      </form>
    )
  }
}
