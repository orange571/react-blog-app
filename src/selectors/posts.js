import moment from 'moment';

// Get visible posts

export default (posts, { text, sortBy }) => {
  return posts.filter((post) => {
    const createdAtMoment = moment(post.createdAt);
    const textMatch = post.title.toLowerCase().includes(text.toLowerCase());

    return textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'title') {
      return a.title.localeCompare(b.title);
    }
  });
};
