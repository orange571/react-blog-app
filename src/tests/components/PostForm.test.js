import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import PostForm from '../../components/PostForm';
import posts from '../fixtures/posts';

test('should render PostForm correctly', () => {
  const wrapper = shallow(<PostForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render PostForm correctly with post data', () => {
  const wrapper = shallow(<PostForm post={posts[1]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
  const wrapper = shallow(<PostForm />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should set title on input change', () => {
  const value = 'New Title';
  const wrapper = shallow(<PostForm />);
  wrapper.find('input').at(0).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('title')).toBe(value);
});

test('should set body on textarea change', () => {
  const value = 'New body value';
  const wrapper = shallow(<PostForm />);
  wrapper.find('textarea').simulate('change', {
    target: { value }
  });
  expect(wrapper.state('body')).toBe(value);
});


test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(<PostForm post={posts[0]} onSubmit={onSubmitSpy} />);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error')).toBe('');
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    title: posts[0].title,
    body: posts[0].body,
    createdAt: posts[0].createdAt,
    author: posts[0].author,
    uid: posts[0].uid,
    isReadable: posts[0].isReadable
  });
});

test('should call onRemove', () => {
  const onRemoveSpy = jest.fn();
  const wrapper = shallow(<PostForm post={posts[0]} onRemove={onRemoveSpy} />);
  wrapper.find('button').at(1).simulate('click');
  expect(onRemoveSpy).toHaveBeenCalled();
});



