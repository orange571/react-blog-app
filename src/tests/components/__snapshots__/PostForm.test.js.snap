// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PostForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a body for your post"
    value=""
  />
  <div>
    <input
      checked={false}
      id="toggle-public-input"
      name="isReadable"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      htmlFor="toggle-public-input"
    >
       Make Public? Your Google display name will be associated with this post.
    </label>
  </div>
  <div
    className="form__button-container"
  >
    <button
      className="button save"
      type="submit"
    >
      Save Post
    </button>
  </div>
</form>
`;

exports[`should render PostForm correctly with post data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value="Rent"
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a body for your post"
    value="Rent is expensive"
  />
  <div>
    <input
      checked={true}
      id="toggle-public-input"
      name="isReadable"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      htmlFor="toggle-public-input"
    >
       Make Public? Your Google display name will be associated with this post.
    </label>
  </div>
  <div
    className="form__button-container"
  >
    <button
      className="button save"
      type="submit"
    >
      Save Post
    </button>
  </div>
</form>
`;

exports[`should render error for invalid form submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a body for your post"
    value=""
  />
  <div>
    <input
      checked={false}
      id="toggle-public-input"
      name="isReadable"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      htmlFor="toggle-public-input"
    >
       Make Public? Your Google display name will be associated with this post.
    </label>
  </div>
  <div
    className="form__button-container"
  >
    <button
      className="button save"
      type="submit"
    >
      Save Post
    </button>
  </div>
</form>
`;

exports[`should render error for invalid form submission 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please provide a title and post body.
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a body for your post"
    value=""
  />
  <div>
    <input
      checked={false}
      id="toggle-public-input"
      name="isReadable"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      htmlFor="toggle-public-input"
    >
       Make Public? Your Google display name will be associated with this post.
    </label>
  </div>
  <div
    className="form__button-container"
  >
    <button
      className="button save"
      type="submit"
    >
      Save Post
    </button>
  </div>
</form>
`;
